### DISCUSSION TITLE
i2s_mcux_sai: noisy sound with k_mem_slab buffer in external SDRAM

### DISCUSSION THREAD

**Discussion by blemouzy** [zephyrproject-rtos/zephyr#61966] [Q&A] [answered]
Hello,

I'm using MCUX SAI audio in TDM mode on a custom i.MX RT1052 board and everything looks good except
that sound is noisy (crackling).
However, the sound is great when k_mem_slab buffer is put in DTCM (at about 500 MHz).

I don't know if this issue is related to:
- MCU (limitation or misconfiguration)
- board (SAI or SDRAM misconfiguration)
- Zephyr (limitation or misconfiguration)
- ...

Has anyone ever had this kind of problem or have any idea to fix this issue?

**Environnement**

- Custom i.MX RT1052 board at about 500 MHz
- External SDRAM (SEMC) at about 50 MHz
- I2S MCLK = 11289750 Hz (11 MHz)

Audio PLL was changed with the following values:
```c
/* Audio PLL Output Frequency is determined by:
 * (Fref * (DIV_SELECT + NUM/DENOM)) / POST_DIV
 * = (24MHz * (30 + 53 / 500)) / 1 = 722.544 MHz
 *
 * PLL is changed here to be able to generate a
 * 44.1 kHz FSYNC instead of 48 kHz one.
 */
pll-clocks = <&anatop 0x70 0xC000 0>,
		<&anatop 0x70 0x7F 30>,
		<&anatop 0x70 0x180000 1>,
		<&anatop 0x80 0x3FFFFFFF 53>,
		<&anatop 0x90 0x3FFFFFFF 500>;
```
but the issue is still the same with the original one.

- Zephyr v3.4.0
- Zephyr default SAI configuration:

```conf
CONFIG_I2S_MCUX_SAI=y
CONFIG_I2S_RX_BLOCK_COUNT=4
CONFIG_I2S_TX_BLOCK_COUNT=4
CONFIG_I2S_EDMA_BURST_SIZE=2
```

- SAI in TDM mode
  - audio sample frequency = 44.1 kHz
  - TDM slot size = 16 bits
  - TDM slot number = 8
  ==> SCLK = 5644800 Hz (5.6 MHz)
- 4 TAS6422 DACs / amplifiers

**Code**

The source code is quiet simple (inspired from `samples/drivers/i2s/echo/src/main.c`) with the
following steps:

```mermaid
sequenceDiagram
    autonumber

    participant codec as Audio codec
    participant main as Main
    participant tdm as SAI TDM
    participant mmc as eMMC

    main->>mmc: Initialize the eMMC (sd_init)
    main->>codec: Configure audio codecs (audio_codec_configure)
    main->>codec: Start audio codecs (audio_codec_start_output)
    main->>tdm: Configure audio bus (i2s_configure)
    main->>main: Get free audio buffer (k_mem_slab_alloc)
    main->>main: Fill the audio buffer with zero
    main->>tdm: Write the audio buffer (i2s_write)
    main->>tdm: Start transmission (i2s_trigger)
    loop
        main->>main: Get free audio buffer (k_mem_slab_alloc)
        main->>mmc: Read data from eMMC (mmc_read_blocks)
        main->>main: Create TDM frames in audio buffer
        main->>tdm: Write the audio buffer (i2s_write)
    end
```

with following defines:

```c
/*
 * Audio parameters
 */
#define AUDIO_SAMPLE_FREQ      (44100)
#define AUDIO_NUM_CHANNELS     (8)
#define AUDIO_SAMPLE_BYTES     (2)
#define AUDIO_SAMPLE_BIT_WIDTH (16)
#define AUDIO_FRAME_BYTES      (AUDIO_NUM_CHANNELS * AUDIO_SAMPLE_BYTES)

/*
 * Audio buffers parameters
 */
#define MMC_SECTOR_COUNT     (4)
#define MMC_SECTOR_SIZE      (512)
#define AUDIO_FRAME_BUF_SIZE (AUDIO_NUM_CHANNELS * MMC_SECTOR_COUNT * MMC_SECTOR_SIZE)
#define I2S_BUF_COUNT        (4)
```

Audio buffers can be placed in external SDRAM (at about 50 MHz):

```c
/* Audio buffers */
static char audio_buffers[I2S_BUF_COUNT * AUDIO_FRAME_BUF_SIZE] __aligned(4);

/* Audio buffers pool */
static struct k_mem_slab audio_buffers_mem_slab;

/* Initialize the the audio buffers memory slab */
k_mem_slab_init(&audio_buffers_mem_slab, audio_buffers, AUDIO_FRAME_BUF_SIZE, I2S_BUF_COUNT);
```

or in internal DTCM (at about 500 MHz):

```c
/* Audio buffers */
static char audio_buffers[I2S_BUF_COUNT * AUDIO_FRAME_BUF_SIZE] __aligned(4) __dtcm_noinit_section;

/* Audio buffers pool */
static struct k_mem_slab audio_buffers_mem_slab;

/* Initialize the the audio buffers memory slab */
k_mem_slab_init(&audio_buffers_mem_slab, audio_buffers, AUDIO_FRAME_BUF_SIZE, I2S_BUF_COUNT);
```

**Issue description**

&#10060; The sound is noisy (crackling) when audio buffers are placed in external SDRAM
&#9989; The sound is fine when audio buffers are placed in internal DTCM

Using a logical analyzer with known pattern (0x00 to 0xFF for MSB and LSB with no eMMC read) shows that when audio
buffers are in external SDRAM, some 0x00 are present between data which explains crackling:

1. Audio buffers in internal DTCM: everything is working at expected
![dtcm](https://github.com/zephyrproject-rtos/zephyr/assets/124877155/ef123aa1-afdb-464c-abce-3018ad719f2e)

2. Audio buffers in external SDRAM: sometimes, data is moved from one frame
![sdram](https://github.com/zephyrproject-rtos/zephyr/assets/124877155/4c7ecaa4-56f4-4353-9b28-4bed3b3fa13d)

3. Audio buffers in external SDRAM: sometimes, data is set to 0
![sdram2](https://github.com/zephyrproject-rtos/zephyr/assets/124877155/0bba5fd0-17ca-41b4-b56c-99851448c500)

It is like SAI DMA is too fast for external SDRAM but I don't think that is really the case
(Ethernet, which also use DMA, is working great at about 70 Mbps).

**Comment by DerekSnell** [comment, marked as answer]
Hi @blemouzy ,
When I see issues like this where things work great from internal non-cached TCMs, but not from cached external memory, my initial thoughts are either a performance bottleneck with the external memory, or issues related to cache.

Reading your description, are you managing cache coherency with the DMA?    When declaring the buffers like this for SDRAM:
```c
/* Audio buffers */
static char audio_buffers[I2S_BUF_COUNT * AUDIO_FRAME_BUF_SIZE] __aligned(4);
```
I believe by default that SDRAM is cached.  In the i.MX RT1xxx MCU architecture, the caches are for the Cortex-M7 core.  Other masters in the system do not access through the cache.  Which means when the DMA reads the SDRAM, it may not be the same data written to the SDRAM by the M7 core.

One option is to place your audio buffers in non-cacheable memory.  You can select [CONFIG_NOCACHE_MEMORY](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_NOCACHE_MEMORY)=y to enable this.  And then use the `__nocache` attribute when you declare the buffers.  

Another option if you wanted to keep these buffers cached, is to manage the cache coherency in your app.  That means after the M7 core writes the buffers, and before the DMA reads the buffers, flush the cache by calling [DCACHE_CleanByRange()](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk/drivers/cache/armv7-m7/fsl_cache.h#L439C6-L439C25) in the MCUXpresso HAL.  And if the DMA writes to the buffers, before the M7 core reads the buffers, call [DCACHE_InvalidateByRange()](https://github.com/zephyrproject-rtos/hal_nxp/blob/b0274481021b6343740f1970ef167d65ea67d30b/mcux/mcux-sdk/drivers/cache/armv7-m7/fsl_cache.h#L425).

Does that help any?

**Comment by blemouzy** [comment]
Hi @DerekSnell ,

> Does that help any?

Oh yes, it does.
This is indeed the root cause of this crackling sound and adding `__nocache` fix it.

Thanks a lot!

**Comment by blemouzy** [comment]
Hello,

I moved my audio sample to Zephyr v3.5.0 and noisy sound is back.
It looks like its still linked to memory caching.

Several new commits are linked to memory / external SDRAM (c3dfc2220a28b4b6b3dc1b782e409780531a52f0,  7e646b56a10f7eb14ba8943a4cb015ad1ee43773, c85d3dd8285aaebb1d3d7043cab503ec2855cb38) but I don't understand what I have to do to make `__nocache` work again.

Is there something to add to board device tree?

**Reply by DerekSnell** [reply]
Hi @blemouzy ,
Did you already add `CONFIG_NXP_IMX_EXTERNAL_SDRAM=y` to your board for v3.5?  You highlighted this commit https://github.com/zephyrproject-rtos/zephyr/commit/7e646b56a10f7eb14ba8943a4cb015ad1ee43773 which introduced that Kconfig.  And the [migration-guide-3.5](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-3.5.rst) states:
> NXP i.MX RT based boards should now enable [:kconfig:option:`CONFIG_DEVICE_CONFIGURATION_DATA`](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-3.5.rst#id91) at the board level when using a DCD with the RT bootrom, and enable [:kconfig:option:`CONFIG_NXP_IMX_EXTERNAL_SDRAM`](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-3.5.rst#id93) when using external SDRAM via the SEMC

If `CONFIG_NXP_IMX_EXTERNAL_SDRAM` is not selected, then the cache is not enabled for the SDRAM.  And perhaps that reduced performance in your app is leading to the issues you see.

Let us know what you find

**Reply by blemouzy** [reply]
Hi @DerekSnell,

Yes, these options are selected (software don't start without them anyway).

After investigation, this issue is linked to https://github.com/zephyrproject-rtos/zephyr/commit/a6adaedd51ffe1b658304da798892bce0a67382b PCM fix.
I have to add `I2S_FMT_CLK_NF_IB` now to have the right clock polarity for TDM.

Thanks for help!
